File: Sources/Bifrost/Services/IndexAutoRefresher.swift
import Foundation

/// On app open, if the cached Thunderstore package index is missing or
/// older than a day, kicks off a quiet background refresh via
/// `ThunderstoreClient`'s existing force-fetch path — no modal, no blocking
/// of startup. `InstalledModsView` already does its own conditional
/// (cheap, 304-friendly) fetch on open; this is purely about not letting the
/// cache go stale for days at a time if Bifrost is left running or is
/// reopened frequently without ever visiting Installed/Browse.
///
/// Reads the cache file's own modification date directly, rather than
/// adding a new accessor to `ThunderstoreClient` — the cache path is a
/// fixed, already-relied-upon convention (see `ThunderstoreClient.init` and
/// `DebugCheck.checkThunderstoreIndex`).
enum IndexAutoRefresher {
    static let staleAfter: TimeInterval = 24 * 60 * 60

    static var cacheFileURL: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Application Support/Bifrost/package-index.json")
    }

    /// True when the cache file is missing entirely, or its modification
    /// date is more than `staleAfter` seconds before `now`.
    static func isStale(now: Date = Date()) -> Bool {
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: cacheFileURL.path),
              let modified = attributes[.modificationDate] as? Date else {
            return true
        }
        return now.timeIntervalSince(modified) > staleAfter
    }

    /// If the cache is stale, kicks off a background force-refresh and
    /// returns the `Task` driving it (so a caller can await it if it wants
    /// to, e.g. to know when to clear a status note); returns `nil` without
    /// calling `onStatus` at all if the cache is already fresh. `onStatus`
    /// is called exactly once, with a short human-readable note, once the
    /// refresh finishes (successfully or not).
    @discardableResult
    static func refreshIfStale(
        client: ThunderstoreClient,
        onStatus: @escaping @Sendable (String) -> Void = { _ in }
    ) -> Task<Void, Never>? {
        guard isStale() else { return nil }
        return Task {
            do {
                let packages = try await client.fetchIndex(force: true)
                onStatus("Thunderstore index refreshed in the background (\(packages.count) packages)")
            } catch {
                onStatus("Background index refresh failed: \(error.localizedDescription)")
            }
        }
    }
}

File: windows/src/Bifrost.Core/Services/IndexAutoRefresher.cs
namespace Bifrost.Core.Services;

/// <summary>
/// On app open, if the cached Thunderstore package index is missing or older
/// than a day, kicks off a quiet background refresh via
/// <see cref="ThunderstoreClient"/>'s existing force-fetch path — no modal,
/// no blocking of startup. Ported from the macOS reference implementation's
/// <c>IndexAutoRefresher.swift</c>.
///
/// Reads the cache file's own last-write time directly, rather than adding a
/// new accessor to <see cref="ThunderstoreClient"/> — the cache path is a
/// fixed, already-relied-upon convention (see <see cref="BifrostPaths.PackageIndexCachePath"/>).
/// </summary>
public static class IndexAutoRefresher
{
    public static readonly TimeSpan StaleAfter = TimeSpan.FromHours(24);

    /// <summary>True when the cache file is missing entirely, or its last-write time is more than <see cref="StaleAfter"/> before <paramref name="now"/> (UTC).</summary>
    public static bool IsStale(string? cacheFilePath = null, DateTime? now = null)
    {
        cacheFilePath ??= BifrostPaths.PackageIndexCachePath;
        if (!File.Exists(cacheFilePath))
        {
            return true;
        }
        var modified = File.GetLastWriteTimeUtc(cacheFilePath);
        return (now ?? DateTime.UtcNow) - modified > StaleAfter;
    }

    /// <summary>
    /// If the cache is stale, kicks off a background force-refresh and
    /// returns a short human-readable status note once it finishes
    /// (successfully or not); returns null without doing anything if the
    /// cache is already fresh.
    /// </summary>
    public static async Task<string?> RefreshIfStaleAsync(ThunderstoreClient client, string? cacheFilePath = null)
    {
        if (!IsStale(cacheFilePath))
        {
            return null;
        }
        try
        {
            var packages = await client.FetchIndexAsync(force: true);
            return $"Thunderstore index refreshed in the background ({packages.Count} packages)";
        }
        catch (Exception ex)
        {
            return $"Background index refresh failed: {ex.Message}";
        }
    }
}
